src/controller.ts

```diff
@@ -1,12 +1,90 @@
-// tslint:disable-next-line:no-empty-interface
-export interface IController {}
+import Optional from "optional.js";
+import { EventHandler, Failure, makePromise, Resolver, Success } from "./events";
+import { INativeView } from "./nativeview";
+import Navigation from "./navigation";
 
-export interface ControllerCreator<T extends IController> {
-    new(...args: any[]): T;
+export interface IController {
+    readonly view: INativeView;
+    getValues(fields: string[]): Promise<string[]>;
+    render(model?: string);
+    renderComponent(name: string, model?: string);
+    display();
+    onBind(): Map<string, EventHandler>;
+    onFinish();
+    onPageCreated();
+    onPageRestart();
+    onPageVisible();
 }
 
-export function createController<T extends IController>(
-    ctor: ControllerCreator<T>,
-    ...args: any[]): T {
-    return new ctor(...args);
+export class BaseController implements IController {
+    public readonly view: INativeView;
+    constructor(view: INativeView, ...args: any[]) {
+        this.view = view;
+    }
+
+    public getValues(fields: string[]): Promise<string[]> {
+        return makePromise((success: Success<string[]>, failure: Failure) => {
+            const wrappedSuccess = (data: string): void => {
+                success(JSON.parse(data));
+            };
+            this.view.getValues(fields, wrappedSuccess, failure );
+        });
+    }
+    public render(model?: string): Promise<{}> {
+        const nullableModel = Optional.ofNullable(model);
+
+        if ( !nullableModel.isPresent() ) {
+            return Promise.reject("Model not present");
+        }
+
+        return nullableModel
+                .map((m) => this.renderOnNativeView(m))
+                .get();
+    }
```

Comment:
I don't see a value of using Nullable for a local variable. As we are wrapping something in it and then immediately checking whether the wrapped thing is null/undefined or not. Also, this is adding complexity as you need to map() before get() when u already know that it's not going to be null (as that guard already exists)

Reply:
Correct. bad sunday refactoring! 🗡